crates/wasi/src/p3/mod.rs

```diff
@@ -79,6 +79,103 @@ where
     }
 }
 
+/// Helper structure to convert an iterator of `Result<T, E>` into a `stream<T>`
+/// plus a `future<result<_, T>>` in WIT.
+///
+/// This will drain the iterator on calls to `poll_produce` and place as many
+/// items as the input buffer has capacity for into the result. This will avoid
+/// doing anything if the async read is cancelled.
+///
+/// Note that this does not actually do anything async, it's assuming that the
+/// internal `iter` is either fast or intended to block.
+struct FallibleIteratorProducer<I, E> {
+    iter: I,
+    result: Option<oneshot::Sender<Result<(), E>>>,
+}
+
+impl<I, T, E, D> StreamProducer<D> for FallibleIteratorProducer<I, E>
+where
+    I: Iterator<Item = Result<T, E>> + Send + Unpin + 'static,
+    T: Send + Sync + 'static,
+    E: Send + 'static,
+{
+    type Item = T;
+    type Buffer = VecBuffer<T>;
+
+    fn poll_produce<'a>(
+        mut self: Pin<&mut Self>,
+        _: &mut Context<'_>,
+        mut store: StoreContextMut<'a, D>,
+        mut dst: Destination<'a, Self::Item, Self::Buffer>,
+        finish: bool,
+    ) -> Poll<wasmtime::Result<StreamResult>> {
+        // Bail out early if this is cancelled.
+        if finish {
+            self.close(Ok(()));
```

Comment:
Is this correct? I was under impression that a *stream.read* can be cancelled, but that does not cancel the underlying stream itself